### DISCUSSION TITLE
The tool does NOT add the needed include's and import's

### DISCUSSION THREAD

**Discussion by dudifrid** [community/community#7388] [Copilot Conversations]
When the tool suggests a snippet of code, it does not suggest also the relevant import's if needed.
I find it frustrating to look for the missing import's, and it seems to be easy to add this ability, as there are so many modern IDE's that allow this ability even without this heavy AI core.

**Edit:** 
Of course that the import's should be added in the right place at the top of the page and not where the cursor is: the place of the cursor makes the tool suggest a snippet of code which, in turn, depends on some import, that in my opinion the tool needs to add if its suggested snippet of code has been accepted by the user.

**Comment by Urethramancer** [comment]
When writing in Go, goimports handles that sort of thing in the background. Other languages are still annoying though - Copilot may have trouble inserting imports in the right context. If Copilot doesn't actually parse a programming language's AST, it just fills in code at the position of the cursor, so an external tool would be necessary. Auto-imports is really a separate thing from AI.

**Reply by dudifrid** [reply]
I edited my question as a repsonse to your comment.

**Comment by iEns** [comment]
In my experience (writing C), CoPilot will suggest random header includes when triggered at the top of the file, seemingly unaware that an error exist because of a missing include.

Suggestions improve immensely when guided by a comment like
`// include header containing MAX_PATH`

**Comment by ctsstc** [comment]
Maybe there could be a post-processing after the autocomplete -- it would have to generate an AST and check if references need to be included which could then leverage current methods/language servers -- if you use standard autocomplete in VSCode it will understand available imports and auto-import when needed. It does seem like the functionality is nearly already there in standard autocomplete 🤞 

## Standard auto-complete auto-imports
![image](https://user-images.githubusercontent.com/416814/141870885-ca4c75a1-8efa-4a28-9341-887cc64a9b6a.png)

Note seems to be a duplicate of: https://github.com/github/feedback/discussions/6975

**Comment by ggoonnzzaallo** [comment]
Noticed the same and pretty frustrated that 2 years+ later this still has not been solved.

I actuall end up building my own Copilot equivalent that handles imports correctly by:

1. Adding the import statement for the right library/function/variable when it's 100% clear
2. Automatically pop-up IntelliSense when there's more than one possible solution

If anyone is interesting to try it, you can download the extension [here](https://marketplace.visualstudio.com/items?itemName=doublebot.doublebot).

**Reply by tendiesfortwo** [reply]
Neat! Will check it out

**Comment by fcFn** [comment]
For me it adds the imports, I think, but not the correct ones 😡 So I have to be constantly on the lookout that Copilot won't insert some rubbish. If anyone knows if there's a reported issue about that, please point me!
